File: ios/TeeTimeCaddie/TeeTimeCaddie/Source/storage/FirebaseFirestoreClient.swift
//
//  FirebaseFirestoreClient.swift
//  TeeTimeCaddie
//
//  Platform (iOS) implementation of the shared FirestoreClient interface, backed by the native
//  Firebase Firestore SDK. Created by the app and passed into the SDK during initialization, so
//  the shared code carries no Firebase dependency. Documents cross the boundary as JSON strings.
//

import Foundation
import TeeTimeCaddieKit
import FirebaseFirestore

class FirebaseFirestoreClient: FirestoreClient {

    private let db: Firestore

    init(useEmulator: Bool) {
        let firestore = Firestore.firestore()
        if useEmulator {
            let settings = firestore.settings
            settings.host = "\(Self.emulatorHost):\(Self.emulatorPort)"
            settings.isSSLEnabled = false
            settings.cacheSettings = MemoryCacheSettings()
            firestore.settings = settings
        }
        db = firestore
    }

    // These satisfy the Kotlin `suspend` requirements. SKIE names the underlying protocol
    // requirements with a `__` prefix (the clean `add`/`get`/... async methods are caller-side
    // sugar that SKIE generates as an extension), so the implementation uses the `__` names.
    func __add(collection: String, json: String) async throws -> String {
        let ref = try await db.collection(collection).addDocument(data: json.toFirestoreDict())
        return ref.documentID
    }

    func __set(collection: String, id: String, json: String) async throws {
        try await db.collection(collection).document(id).setData(json.toFirestoreDict())
    }

    func __get(collection: String, id: String) async throws -> StoredDoc? {
        let snapshot = try await db.collection(collection).document(id).getDocument()
        guard snapshot.exists, let data = snapshot.data() else { return nil }
        return StoredDoc(id: snapshot.documentID, json: data.toJsonString())
    }

    func __query(collection: String, whereField: String, whereValue: String, orderByField: String, descending: Bool) async throws -> [StoredDoc] {
        let snapshot = try await db.collection(collection)
            .whereField(whereField, isEqualTo: whereValue)
            .order(by: orderByField, descending: descending)
            .getDocuments()
        return snapshot.documents.map { StoredDoc(id: $0.documentID, json: $0.data().toJsonString()) }
    }

    func observeQuery(
        collection: String,
        whereField: String,
        whereValue: String,
        orderByField: String,
        descending: Bool,
        onChange: @escaping ([StoredDoc]) -> Void,
        onError: @escaping (String) -> Void
    ) -> TeeTimeCaddieKit.Cancellable {
        let registration = db.collection(collection)
            .whereField(whereField, isEqualTo: whereValue)
            .order(by: orderByField, descending: descending)
            .addSnapshotListener { snapshot, error in
                if let error = error {
                    onError(error.localizedDescription)
                } else if let snapshot = snapshot {
                    onChange(snapshot.documents.map { StoredDoc(id: $0.documentID, json: $0.data().toJsonString()) })
                }
            }
        return ListenerCancellable(registration)
    }

    private static let emulatorHost = "127.0.0.1"
    private static let emulatorPort = 9399
}

private class ListenerCancellable: TeeTimeCaddieKit.Cancellable {
    private let registration: ListenerRegistration
    init(_ registration: ListenerRegistration) { self.registration = registration }
    func cancel() { registration.remove() }
}

private extension String {
    /// Parse this JSON object string into a Firestore-compatible dictionary.
    func toFirestoreDict() -> [String: Any] {
        guard let data = data(using: .utf8),
              let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return [:]
        }
        return dict
    }
}

private extension Dictionary where Key == String, Value == Any {
    /// Serialize this Firestore document dictionary into a JSON object string.
    func toJsonString() -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: self),
              let string = String(data: data, encoding: .utf8) else {
            return "{}"
        }
        return string
    }
}

File: android/app/src/main/java/net/bradball/teetimecaddie/android/storage/FirebaseFirestoreClient.kt
package net.bradball.teetimecaddie.android.storage

import com.google.android.gms.tasks.Task
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.firestoreSettings
import com.google.firebase.firestore.memoryCacheSettings
import kotlinx.coroutines.suspendCancellableCoroutine
import net.bradball.teetimecaddie.core.models.storage.Cancellable
import net.bradball.teetimecaddie.core.models.storage.FirestoreClient
import net.bradball.teetimecaddie.core.models.storage.StoredDoc
import org.json.JSONArray
import org.json.JSONObject
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Android [FirestoreClient] implementation backed by the native Firebase Firestore SDK.
 *
 * This is the platform-provided binding that replaces the GitLive Firestore library. Documents
 * cross the shared-code boundary as JSON strings; this class converts between JSON and Firestore's
 * native map representation.
 */
class FirebaseFirestoreClient(useEmulator: Boolean) : FirestoreClient {

    private val db: FirebaseFirestore = FirebaseFirestore.getInstance().apply {
        if (useEmulator) {
            useEmulator(EMULATOR_HOST, EMULATOR_PORT)
            firestoreSettings = firestoreSettings {
                // Disable local persistence against the emulator, matching prior behaviour.
                setLocalCacheSettings(memoryCacheSettings {})
            }
        }
    }

    override suspend fun add(collection: String, json: String): String =
        db.collection(collection).add(json.toFirestoreMap()).awaitResult().id

    override suspend fun set(collection: String, id: String, json: String) {
        db.collection(collection).document(id).set(json.toFirestoreMap()).awaitResult()
    }

    override suspend fun get(collection: String, id: String): StoredDoc? {
        val snapshot = db.collection(collection).document(id).get().awaitResult()
        return if (snapshot.exists()) snapshot.toStoredDoc() else null
    }

    override suspend fun query(
        collection: String,
        whereField: String,
        whereValue: String,
        orderByField: String,
        descending: Boolean
    ): List<StoredDoc> {
        val snapshot = db.collection(collection)
            .whereEqualTo(whereField, whereValue)
            .orderBy(orderByField, direction(descending))
            .get()
            .awaitResult()
        return snapshot.documents.map { it.toStoredDoc() }
    }

    override fun observeQuery(
        collection: String,
        whereField: String,
        whereValue: String,
        orderByField: String,
        descending: Boolean,
        onChange: (List<StoredDoc>) -> Unit,
        onError: (String) -> Unit
    ): Cancellable {
        val registration: ListenerRegistration = db.collection(collection)
            .whereEqualTo(whereField, whereValue)
            .orderBy(orderByField, direction(descending))
            .addSnapshotListener { snapshot, error ->
                when {
                    error != null -> onError(error.message ?: "Unknown Firestore error")
                    snapshot != null -> onChange(snapshot.documents.map { it.toStoredDoc() })
                }
            }
        return object : Cancellable {
            override fun cancel() = registration.remove()
        }
    }

    private fun direction(descending: Boolean): Query.Direction =
        if (descending) Query.Direction.DESCENDING else Query.Direction.ASCENDING

    companion object {
        private const val EMULATOR_HOST = "10.0.2.2"
        private const val EMULATOR_PORT = 9399
    }
}

private fun DocumentSnapshot.toStoredDoc(): StoredDoc =
    StoredDoc(id, JSONObject(data ?: emptyMap<String, Any>()).toString())

private suspend fun <T> Task<T>.awaitResult(): T = suspendCancellableCoroutine { continuation ->
    addOnSuccessListener { result -> if (continuation.isActive) continuation.resume(result) }
    addOnFailureListener { error -> if (continuation.isActive) continuation.resumeWithException(error) }
}

private fun String.toFirestoreMap(): Map<String, Any?> = JSONObject(this).toValueMap()

private fun JSONObject.toValueMap(): Map<String, Any?> = buildMap {
    val jsonKeys = keys()
    while (jsonKeys.hasNext()) {
        val key = jsonKeys.next()
        put(key, get(key).unwrapJson())
    }
}

private fun JSONArray.toValueList(): List<Any?> = (0 until length()).map { get(it).unwrapJson() }

private fun Any?.unwrapJson(): Any? = when (this) {
    is JSONObject -> toValueMap()
    is JSONArray -> toValueList()
    JSONObject.NULL -> null
    else -> this
}
